### PR TITLE
Added rules for required for engine and datacapture libraries.

### LINKED ISSUE #1901: SDC : Proguard rules not available for apk shrink
**Describe the bug**
While applying Proguard rules to shrink apk, app crashes.

**Component**
SDC library - demo reference app

**To Reproduce**
1. Go to build.gradle.kts (:demo) file
2. set debug build type config with "isMinifyEnabled" as true. Add or update following lines in  buildTypes {}
 getByName("debug") {
      isMinifyEnabled = true
      proguardFiles(getDefaultProguardFile("proguard-android-optimize.txt"), "proguard-rules.pro")
    }
     getByName("release") {
      isMinifyEnabled = false
      proguardFiles(getDefaultProguardFile("proguard-android-optimize.txt"), "proguard-rules.pro")
    }
3. sync and run app in device
4. App crashes with an exception : 
 java.util.MissingResourceException: Can't find bundle for base name h2.hapi-messages, locale en_US
        at java.util.ResourceBundle.throwMissingResourceException(ResourceBundle.java:1561)
        at java.util.ResourceBundle.getBundleImpl(ResourceBundle.java:1384)
        at java.util.ResourceBundle.getBundle(ResourceBundle.java:771)
        at h2.a.e(:175)
        at h2.a.<init>(:64)
        at h2.a.<init>(:60)
        at e2.k.<init>(:115)
        at e2.k.<init>(:175)
        at e2.k.v(:1199)
        at e2.k.a(Unknown Source:0)
        at e2.j.apply(Unknown Source:2)
        at java.util.Map.computeIfAbsent(Map.java:1000)
        at java.util.Collections$SynchronizedMap.computeIfAbsent(Collections.java:2723)
        at e2.k.c(:1199)
        at n5.h$a.b(:71)
        at n5.g.d(:77)
        at n5.g.c(:52)
        at com.google.android.fhir.demo.FhirApplication.d(:67)
        at com.google.android.fhir.demo.FhirApplication.b(:32)
        at com.google.android.fhir.demo.FhirApplication$b.c(:34)
        at com.google.android.fhir.demo.FhirApplication$b.g(:34)

**Expected behavior**
Need required Proguard rules to avoid missing resource exception

**Smartphone (please complete the following information):**
 - Device: Pixel 4 emulator
 - Android version: 31
 
 @jingtang10

### PR DESCRIPTION
**Pull request by aditya-07**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #1901 #2327

**Description**
~~Added the proguard rules required for engine and datacapture libs in Demo and catalog app respectively.~~
Added rules to respective `proguard-rules.pro` files for engine and datacapture library as per [Android's considerations](https://developer.android.com/studio/projects/android-library#Considerations).
This way, R8 automatically applies the provided rules and developers don't have to include the rules explicitly in the app's `proguard-rules` file ( see [Library dependencies](https://developer.android.com/studio/build/shrink-code#configuration-files) ).

**Alternative(s) considered**
Publish the rules in wiki, but then users will have to check it with every release to check if rules have been modified.

**Type**
Choose one: Bug fix

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

File: catalog/proguard-rules.pro [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+# Uncomment this to preserve the line number information for
+# debugging stack traces.
+#-keepattributes SourceFile,LineNumberTable
+
+# If you keep the line number information, uncomment this to
+# hide the original source file name.
+#-renamesourcefileattribute SourceFile
+
+-keep class com.google.android.fhir.catalog.WorkflowType
```
Comment (omarismail94) [inline catalog/proguard-rules.pro:23]: nit: add new line

File: engine/proguard-rules.pro [hunk 2]
```diff
@@ -19,3 +19,24 @@
...
+## sqlcipher starts
+# see (https://github.com/sqlcipher/android-database-sqlcipher/tree/master#proguard)
+-keep,includedescriptorclasses class net.sqlcipher.** { *; }
+-keep,includedescriptorclasses interface net.sqlcipher.** { *; }
+## sqlcipher endss
+
+## retrofit starts
+# see (https://github.com/square/retrofit/issues/3539)
+-keep class com.google.android.fhir.sync.remote.RemoteFhirService { *; }
+## retrofit ends
```
Comment (omarismail94) [inline engine/proguard-rules.pro:42]: nit: add new line

**Review by omarismail94 — APPROVED** [review]
Great change @aditya-07 ! Just two nits

**Comment by omarismail94** [comment]
Im also assuming we don't need this in any of our other libraries as they depend on engine and datacapture

File: catalog/proguard-rules.pro [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+#   public *;
+#}
+
+# Uncomment this to preserve the line number information for
+# debugging stack traces.
+#-keepattributes SourceFile,LineNumberTable
+
+# If you keep the line number information, uncomment this to
+# hide the original source file name.
+#-renamesourcefileattribute SourceFile
```
Comment (jingtang10) [inline catalog/proguard-rules.pro:21]: these can be removed. i don't mind if you keep the block on top - up to you.

and below

File: catalog/proguard-rules.pro (hunk 1 above)
Comment (jingtang10) [inline catalog/proguard-rules.pro:23]: i don't understand why this is needed. as a argument to pass to the fragment, i feel this should still work even if it's minified/shrunk?

File: datacapture/proguard-rules.pro [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+# debugging stack traces.
+#-keepattributes SourceFile,LineNumberTable
+
+# If you keep the line number information, uncomment this to
+# hide the original source file name.
+#-renamesourcefileattribute SourceFile
+
+## hapi libs starts
+-keep class ca.uhn.fhir.** { *; }
+-keep class org.hl7.fhir.r4.hapi.ctx.*  { *; }
```
Comment (jingtang10) [inline datacapture/proguard-rules.pro:25]: can you do `org.hl7.fhir.*` ?

File: datacapture/proguard-rules.pro [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+-keep class org.hl7.fhir.r4.**  { *; }
+-keep class org.hl7.fhir.utilities.**  { *; }
+-keep class org.hl7.fhir.exceptions.*  { *; }
```
Comment (jingtang10) [inline datacapture/proguard-rules.pro:28]: why are these needed?

File: datacapture/proguard-rules.pro [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+# Uncomment this to preserve the line number information for
+# debugging stack traces.
+#-keepattributes SourceFile,LineNumberTable
+
+# If you keep the line number information, uncomment this to
+# hide the original source file name.
+#-renamesourcefileattribute SourceFile
+
+## hapi libs starts
```
Comment (jingtang10) [inline datacapture/proguard-rules.pro:23]:
```suggestion
# HAPI Strucutres need to be kept beause
# 1.  Reflection is used in resource extraction as the FHIR Path expression is then translated to field names to locate data elements in FHIR resources
# 2. In HAPI Strucures, ClassLoader is used...
```

**Review by jingtang10 — CHANGES_REQUESTED** [review]

File: catalog/proguard-rules.pro (hunk 1 above)
Comment (aditya-07) [reply catalog/proguard-rules.pro:23]: Done

File: engine/proguard-rules.pro (hunk 2 above)
Comment (aditya-07) [reply engine/proguard-rules.pro]: done

File: catalog/proguard-rules.pro (hunk 1 above)
Comment (aditya-07) [reply catalog/proguard-rules.pro:23]: Removed as not required in catalog app anymore.

Comment (aditya-07) [reply catalog/proguard-rules.pro:23]: Even if something similar is required later, it needs to go in Catalog's proguard-rules and not here.

File: datacapture/proguard-rules.pro (hunk 5 above)
Comment (aditya-07) [reply datacapture/proguard-rules.pro]: removed. Keeping the top package now

File: datacapture/proguard-rules.pro (hunk 6 above)
Comment (aditya-07) [reply datacapture/proguard-rules.pro]: done

File: catalog/proguard-rules.pro (hunk 1 above)
Comment (jingtang10) [inline catalog/proguard-rules.pro:23]: you can't delete this?

File: datacapture/proguard-rules.pro
Comment (jingtang10) [inline datacapture/proguard-rules.pro:1]: this is excellent - thanks so much for updating the doc!

File: catalog/proguard-rules.pro (hunk 1 above)
Comment (jingtang10) [reply catalog/proguard-rules.pro:23]: i actually just built a release apk of the catalog app without this rule and it seems to be working...

so can you just remove this file altogether

File: catalog/build.gradle.kts [hunk 7]
```diff
@@ -22,8 +22,8 @@ android {
 
   buildTypes {
     release {
-      isMinifyEnabled = false
-      proguardFiles(getDefaultProguardFile("proguard-android-optimize.txt"))
+      isMinifyEnabled = true
```
Comment (jingtang10) [inline catalog/build.gradle.kts:25]: and demo app?

**Review by jingtang10 — CHANGES_REQUESTED** [review]

File: catalog/build.gradle.kts (hunk 7 above)
Comment (aditya-07) [reply catalog/build.gradle.kts:25]: its there

**Comment by jingtang10** [comment]
tested a minified release build of the demo app and got this:

```
2024-06-13 17:22:21.350 27518-27518 AndroidRuntime          com.google.android.fhir.demo         E  FATAL EXCEPTION: main
                                                                                                    Process: com.google.android.fhir.demo, PID: 27518
                                                                                                    ca.uhn.fhir.context.ConfigurationException: HAPI-1753: Unable to initialize StAX - XML processing is disabled
                                                                                                    	at ca.uhn.fhir.util.XmlUtil.newInputFactory(SourceFile:59)
                                                                                                    	at ca.uhn.fhir.util.XmlUtil.getOrCreateInputFactory(SourceFile:1)
                                                                                                    	at ca.uhn.fhir.util.XmlUtil.createXmlReader(SourceFile:1)
                                                                                                    	at ca.uhn.fhir.parser.XmlParser.createStreamReader(SourceFile:1)
                                                                                                    	at ca.uhn.fhir.parser.XmlParser.doParseResource(Unknown Source:0)
                                                                                                    	at ca.uhn.fhir.parser.BaseParser.parseResource(SourceFile:4)
                                                                                                    	at ca.uhn.fhir.parser.BaseParser.parseResource(SourceFile:2)
                                                                                                    	at ca.uhn.fhir.context.support.DefaultProfileValidationSupportBundleStrategy.parseBundle(Unknown Source:8)
                                                                                                    	at ca.uhn.fhir.context.support.DefaultProfileValidationSupportBundleStrategy.loadStructureDefinitions(SourceFile:80)
                                                                                                    	at ca.uhn.fhir.context.support.DefaultProfileValidationSupportBundleStrategy.provideStructureDefinitionMap(Unknown Source:30)
                                                                                                    	at ca.uhn.fhir.context.support.DefaultProfileValidationSupportBundleStrategy.fetchStructureDefinition(SourceFile:62)
                                                                                                    	at ca.uhn.fhir.context.support.DefaultProfileValidationSupport.fetchStructureDefinition(Unknown Source:2)
                                                                                                    	at ca.uhn.fhir.context.support.IValidationSupport.fetchResource(Unknown Source:136)
                                                                                                    	at org.hl7.fhir.common.hapi.validation.support.ValidationSupportChain.fetchResource(Unknown Source:18)
                                                                                                    	at org.hl7.fhir.r4.hapi.ctx.HapiWorkerContext.lambda$fetchResource$0(Unknown Source:2)
                                                                                                    	at org.hl7.fhir.r4.hapi.ctx.HapiWorkerContext.b(SourceFile:1)
                                                                                                    	at org.hl7.fhir.common.hapi.validation.support.f.apply(SourceFile:1)
                                                                                                    	at ca.uhn.fhir.sl.cache.guava.CacheDelegator.lambda$get$0(Unknown Source:0)
                                                                                                    	at ca.uhn.fhir.sl.cache.guava.CacheDelegator.a(SourceFile:1)
                                                                                                    	at ca.uhn.fhir.sl.cache.guava.a.call(SourceFile:1)
                                                                                                    	at a8.j0.load(SourceFile:1)
                                                                                                    	at a8.h0.h(SourceFile:28)
                                                                                                    	at a8.m0.l(SourceFile:185)
                                                                                                    	at a8.b1.e(SourceFile:74)
                                                                                                    	at h4.t.b(SourceFile:1)
                                                                                                    	at ca.uhn.fhir.sl.cache.guava.CacheDelegator.get(SourceFile:10)
                                                                                                    	at org.hl7.fhir.r4.hapi.ctx.HapiWorkerContext.fetchResource(Unknown Source:15)
                                                                                                    	at org.hl7.fhir.r4.hapi.ctx.HapiWorkerContext.fetchResourceWithException(SourceFile:1)
                                                                                                    	at i6.a.fetchResourceWithException(SourceFile:1)
                                                                                                    	at org.hl7.fhir.r4.context.BaseWorkerContext.fetchResource(Unknown Source:0)
                                                                                                    	at org.hl7.fhir.r4.context.SimpleWorkerContext.fetchResource(SourceFile:1)
                                                                                                    	at org.hl7.fhir.r4.context.BaseWorkerContext.fetchTypeDefinition(Unknown Source:8)
                                                                                                    	at org.hl7.fhir.r4.utils.FHIRPathEngine.execute(SourceFile:2)
                                                                                                    	at org.hl7.fhir.r4.utils.FHIRPathEngine.execute(SourceFile:1)
                                                                                                    	at org.hl7.fhir.r4.utils.FHIRPathEngine.evaluate(SourceFile:4)
                                                                                                    	at i6.d.a(SourceFile:7)
                                                                                                    	at b6.d0.g(SourceFile:220)
                                                                                                    	at a6.i.r(SourceFile:134)
                                                                                                    	at a6.i.h(SourceFile:16)
                                                                                                    	at m1.c0.r(SourceFile:68)
                                                                                                    	at m1.c0.m(SourceFile:13)
                                                                                                    	at s3.u0.A0(SourceFile:5)
                                                                                                    	at g8.b.u1(SourceFile:75)
                                                                                                    	at m1.b0.r(SourceFile:88)
                                                                                                    	at l9.a.f(SourceFile:1)
                                                                                                    	at aa.b0.run(SourceFile:1)
                                                                                                    	at aa.i0.P(SourceFile:159)
                                                                                                    	at g8.b.c1(SourceFile:99)
                                                                                                    	at g.g.run(SourceFile:169)
                                                                                                    	at d.p0.run(SourceFile:179)
                                                                                                    	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
                                                                                                    	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:644)
                                                                                                    	at java.lang.Thread.run(Thread.java:1012)
                                                                                                    	Suppressed: fa.f: [i1{Cancelling}@2d8872b, Dispatchers.Main.immediate]
                                                                                                    Caused by: z8.a: Provider com.ctc.wstx.stax.WstxInputFactory not found
                                                                                                    	at ca.uhn.fhir.util.XmlUtil.newInputFactory(SourceFile:41)
                                                                                                    	... 52 more
```

**Review by jingtang10 — APPROVED** [review]

**Comment by aditya-07** [comment]
> tested a minified release build of the demo app and got this when i create a patient:
> 
> ```
> 2024-06-13 17:22:21.350 27518-27518 AndroidRuntime          com.google.android.fhir.demo         E  FATAL EXCEPTION: main
>                                                                                                     Process: com.google.android.fhir.demo, PID: 27518
>                                                                                                     ca.uhn.fhir.context.ConfigurationException: HAPI-1753: Unable to initialize StAX - XML processing is disabled
>                                                                                                     	at ca.uhn.fhir.util.XmlUtil.newInputFactory(SourceFile:59)
>                                                                                                     	at ca.uhn.fhir.util.XmlUtil.getOrCreateInputFactory(SourceFile:1)
>                                                                                                     	at ca.uhn.fhir.util.XmlUtil.createXmlReader(SourceFile:1)
>                                                                                                     	at ca.uhn.fhir.parser.XmlParser.createStreamReader(SourceFile:1)
>                                                                                                     	at ca.uhn.fhir.parser.XmlParser.doParseResource(Unknown Source:0)
>                                                                                                     	at ca.uhn.fhir.parser.BaseParser.parseResource(SourceFile:4)
>                                                                                                     	at ca.uhn.fhir.parser.BaseParser.parseResource(SourceFile:2)
>                                                                                                     	at ca.uhn.fhir.context.support.DefaultProfileValidationSupportBundleStrategy.parseBundle(Unknown Source:8)
>                                                                                                     	at ca.uhn.fhir.context.support.DefaultProfileValidationSupportBundleStrategy.loadStructureDefinitions(SourceFile:80)
>                                                                                                     	at ca.uhn.fhir.context.support.DefaultProfileValidationSupportBundleStrategy.provideStructureDefinitionMap(Unknown Source:30)
>                                                                                                     	at ca.uhn.fhir.context.support.DefaultProfileValidationSupportBundleStrategy.fetchStructureDefinition(SourceFile:62)
>                                                                                                     	at ca.uhn.fhir.context.support.DefaultProfileValidationSupport.fetchStructureDefinition(Unknown Source:2)
>                                                                                                     	at ca.uhn.fhir.context.support.IValidationSupport.fetchResource(Unknown Source:136)
>                                                                                                     	at org.hl7.fhir.common.hapi.validation.support.ValidationSupportChain.fetchResource(Unknown Source:18)
>                                                                                                     	at org.hl7.fhir.r4.hapi.ctx.HapiWorkerContext.lambda$fetchResource$0(Unknown Source:2)
>                                                                                                     	at org.hl7.fhir.r4.hapi.ctx.HapiWorkerContext.b(SourceFile:1)
>                                                                                                     	at org.hl7.fhir.common.hapi.validation.support.f.apply(SourceFile:1)
>                                                                                                     	at ca.uhn.fhir.sl.cache.guava.CacheDelegator.lambda$get$0(Unknown Source:0)
>                                                                                                     	at ca.uhn.fhir.sl.cache.guava.CacheDelegator.a(SourceFile:1)
>                                                                                                     	at ca.uhn.fhir.sl.cache.guava.a.call(SourceFile:1)
>                                                                                                     	at a8.j0.load(SourceFile:1)
>                                                                                                     	at a8.h0.h(SourceFile:28)
>                                                                                                     	at a8.m0.l(SourceFile:185)
>                                                                                                     	at a8.b1.e(SourceFile:74)
>                                                                                                     	at h4.t.b(SourceFile:1)
>                                                                                                     	at ca.uhn.fhir.sl.cache.guava.CacheDelegator.get(SourceFile:10)
>                                                                                                     	at org.hl7.fhir.r4.hapi.ctx.HapiWorkerContext.fetchResource(Unknown Source:15)
>                                                                                                     	at org.hl7.fhir.r4.hapi.ctx.HapiWorkerContext.fetchResourceWithException(SourceFile:1)
>                                                                                                     	at i6.a.fetchResourceWithException(SourceFile:1)
>                                                                                                     	at org.hl7.fhir.r4.context.BaseWorkerContext.fetchResource(Unknown Source:0)
>                                                                                                     	at org.hl7.fhir.r4.context.SimpleWorkerContext.fetchResource(SourceFile:1)
>                                                                                                     	at org.hl7.fhir.r4.context.BaseWorkerContext.fetchTypeDefinition(Unknown Source:8)
>                                                                                                     	at org.hl7.fhir.r4.utils.FHIRPathEngine.execute(SourceFile:2)
>                                                                                                     	at org.hl7.fhir.r4.utils.FHIRPathEngine.execute(SourceFile:1)
>                                                                                                     	at org.hl7.fhir.r4.utils.FHIRPathEngine.evaluate(SourceFile:4)
>                                                                                                     	at i6.d.a(SourceFile:7)
>                                                                                                     	at b6.d0.g(SourceFile:220)
>                                                                                                     	at a6.i.r(SourceFile:134)
>                                                                                                     	at a6.i.h(SourceFile:16)
>                                                                                                     	at m1.c0.r(SourceFile:68)
>                                                                                                     	at m1.c0.m(SourceFile:13)
>                                                                                                     	at s3.u0.A0(SourceFile:5)
>                                                                                                     	at g8.b.u1(SourceFile:75)
>                                                                                                     	at m1.b0.r(SourceFile:88)
>                                                                                                     	at l9.a.f(SourceFile:1)
>                                                                                                     	at aa.b0.run(SourceFile:1)
>                                                                                                     	at aa.i0.P(SourceFile:159)
>                                                                                                     	at g8.b.c1(SourceFile:99)
>                                                                                                     	at g.g.run(SourceFile:169)
>                                                                                                     	at d.p0.run(SourceFile:179)
>                                                                                                     	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
>                                                                                                     	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:644)
>                                                                                                     	at java.lang.Thread.run(Thread.java:1012)
>                                                                                                     	Suppressed: fa.f: [i1{Cancelling}@2d8872b, Dispatchers.Main.immediate]
>                                                                                                     Caused by: z8.a: Provider com.ctc.wstx.stax.WstxInputFactory not found
>                                                                                                     	at ca.uhn.fhir.util.XmlUtil.newInputFactory(SourceFile:41)
>                                                                                                     	... 52 more
> ```

Its fixed now with new rules update.

**Comment by jingtang10** [comment]
Thanks Aditya!

**Pull request merged**